### PR TITLE
Pull request: Node/Link Transport

### PR DESCRIPTION
**Pull request by jonaskirner**
Pull Request: Node/Link Transport

1. Creating paths along nodes for transportressources (through linktransportprocesses) -> Enables fixed routing points
2. Specific routes need specific transportresources (through requiredtransportprocesses) -> Enables conveyor

### CODE REVIEW

File: prodsys/adapters/json_adapter.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ def read_data_old(self, file_path: str, scenario_file_path: Optional[str] = None
         self.queue_data = self.create_objects_from_configuration_data_old(data["queues"], queue_data.QueueData)
         self.resource_data = self.create_objects_from_configuration_data_old(data["resources"], resource_data.RESOURCE_DATA_UNION)
         self.product_data = self.create_objects_from_configuration_data_old(data["products"], product_data.ProductData)
+        # self.link_data = self.create_objects_from_configuration_data(data["links"], links_data.LinkData)
```
Comment (SebBehrendt) [inline prodsys/adapters/json_adapter.py:71]: Remove comment

File: prodsys/adapters/adapter.py [hunk 2]
```diff
@@ -681,6 +682,7 @@ def check_states(cls, state: state_data.STATE_DATA_UNION, values):
             )
         return state
 
+    #TODO: Add validator for nodes_data
```
Comment (SebBehrendt) [inline prodsys/adapters/adapter.py:685]: Add validation

Comment (SebBehrendt) [inline prodsys/adapters/adapter.py:685]: Add validation and Not only comment please

File: prodsys/adapters/adapter.py
Comment (SebBehrendt) [inline prodsys/adapters/adapter.py:1]: The processes available checking should not be avoided with commenting out...

File: prodsys/express/process.py
Comment (SebBehrendt) [inline prodsys/express/process.py:1]: Capabilities should not be optional in capability processes. Also the type hints for links should be explained in more detail in the docstring of the LinkTransportProcess

File: prodsys/factories/linktransportprocess_factory.py
Comment (SebBehrendt) [inline prodsys/factories/linktransportprocess_factory.py:1]: This should not be seperated from process factory -> please integrate it there

File: prodsys/factories/resource_factory.py
Comment (SebBehrendt) [inline prodsys/factories/resource_factory.py:1]: A dedicated node_factory would make more sense, I think. Please extract it to a seperate file.

File: prodsys/simulation/control.py [hunk 3]
```diff
@@ -433,14 +444,30 @@ def start_process(self) -> Generator:
                         ],
                     )
                 logger.debug({"ID": "controller", "sim_time": self.env.now, "resource": self.resource.data.ID, "event": f"Starting picking up {product.product_data.ID} for transport"})
-                yield self.env.process(
-                    self.run_process(transport_state, product, target=origin, empty_transport=True)
-                )
+
+                if isinstance(process, LinkTransportProcess) or isinstance(process, RequiredCapabilityProcess):
```
Comment (SebBehrendt) [inline prodsys/simulation/control.py:448]: This looks a little bit full -> move to extra function or make logging more efficient.

File: prodsys/simulation/control.py [hunk 4]
```diff
@@ -460,9 +487,31 @@ def start_process(self) -> Generator:
                     ],
                 )
             logger.debug({"ID": "controller", "sim_time": self.env.now, "resource": self.resource.data.ID, "event": f"Starting transport of {product.product_data.ID}"})
-            yield self.env.process(
-                self.run_process(transport_state, product, target=target, empty_transport=False)
-            )
+
```
Comment (SebBehrendt) [inline prodsys/simulation/control.py:490]: Same as above

File: prodsys/simulation/control.py [hunk 5]
```diff
@@ -510,9 +570,29 @@ def run_process(
             self.env.now,
             state.StateTypeEnum.transport,
         )
-        input_state.process = self.env.process(
-            input_state.process_state(target=target_location)  # type: ignore False
+        flag = False
```
Comment (SebBehrendt) [inline prodsys/simulation/control.py:573]: make this more elegant, by reusing the same elementary transport function with both normal and link transport

File: prodsys/simulation/path_finder.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+            request (TransportResquest): The transportation request.
+            given_links_list: The given links list.
+
+        Returns:
+            The edges as a list of tuples (with a start_node, end_node and related costs).
+        """
+        from prodsys.simulation import resources, sink, source
+        pathfinder_edges = []
+
+        for link in given_links_list:
```
Comment (SebBehrendt) [inline prodsys/simulation/path_finder.py:60]: Do less nesting here in the code -> move logical bits to separate functions, to make code more readable

File: prodsys/simulation/path_finder.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            given_links_list: The given links list.
+
+        Returns:
+            The origin and target as graph nodes.
+        """
+
+        from prodsys.simulation import resources
+        origin, target = None, None
+
+        for link in given_links_list:
```
Comment (SebBehrendt) [inline prodsys/simulation/path_finder.py:128]: same as above -> fewer nesting and more understandable code

File: prodsys/simulation/process.py [hunk 8]
```diff
@@ -218,42 +224,77 @@ def get_expected_process_time(self) -> float:
...
-        # return True
-        pass
+        if isinstance(requested_process, RequiredCapabilityProcess):
+            if requested_process.process_data.capability == self.process_data.capability:
+                pathfinder = path_finder.Pathfinder()
+                which_path: bool = False
+                path = pathfinder.find_path(request, which_path, self)
+                if not path:
+                    return False
+                else:
```
Comment (SebBehrendt) [inline prodsys/simulation/process.py:245]: remove else, since after return

File: prodsys/simulation/request.py [hunk 9]
```diff
@@ -80,18 +80,35 @@ class TransportResquest(Request):
     """
     def __init__(
         self,
-        process: process.TransportProcess,
+        process: Union[process.TransportProcess, process.LinkTransportProcess],
         product: product.Product,
         origin: product.Location,
         target: product.Location,
+        path = None,
```
Comment (SebBehrendt) [inline prodsys/simulation/request.py:87]: I think this is really inconvenient -> the transport request should hold information about the process to perform, but why here information about the path to take. All of this should be contained in the controller...

File: prodsys/simulation/resources.py [hunk 10]
```diff
@@ -206,11 +222,16 @@ def get_processes(self, process: process.PROCESS_UNION) -> List[state.State]:
         Returns:
             List[state.State]: The state of the resource for the process.
         """
-        possible_states = [
-            actual_state
-            for actual_state in self.production_states
-            if actual_state.state_data.ID == process.process_data.ID
-        ]
+        from prodsys.simulation.process import RequiredCapabilityProcess
```
Comment (SebBehrendt) [inline prodsys/simulation/resources.py:225]: imports at the beginning of the file

File: prodsys/simulation/resources.py [hunk 11]
```diff
@@ -206,11 +222,16 @@ def get_processes(self, process: process.PROCESS_UNION) -> List[state.State]:
...
+        possible_states = []
+        for actual_state in self.production_states:
+            if isinstance(process, RequiredCapabilityProcess) and (
```
Comment (SebBehrendt) [inline prodsys/simulation/resources.py:228]: the capability of the requiredCapabilityProcess should be used, as for all other processes, in the match_request function. Here is logic for process matching -> move it to the process class.

File: prodsys/simulation/resources.py [hunk 12]
```diff
@@ -227,11 +248,14 @@ def get_free_process(self, process: process.PROCESS_UNION) -> Optional[state.Sta
         Returns:
             Optional[state.State]: The state of the resource for the process.
         """
+        from prodsys.simulation.process import RequiredCapabilityProcess
```
Comment (SebBehrendt) [inline prodsys/simulation/resources.py:251]: same as above

File: prodsys/simulation/router.py [hunk 13]
```diff
@@ -54,38 +54,43 @@ def route_request(self, processing_request: request.Request) -> Generator:
...
-                [resource.got_free for resource in possible_resources],
+                [resource.got_free for resource in possible_resources_reached],
             )
             logger.debug({"ID": processing_request.product.product_data.ID, "sim_time": processing_request.product.env.now, "event": f"Free resources available."})
 
 
         if not free_resources:
             raise ValueError("No free resources available, Error in Event handling of routing to resources.")
         routed_resource = free_resources[0]
+        if isinstance(routed_resource, resources.TransportResource) and (isinstance(processing_request.process, process.RequiredCapabilityProcess) or isinstance(processing_request.process, process.LinkTransportProcess)):
```
Comment (SebBehrendt) [inline prodsys/simulation/router.py:89]: This is not very understandable. Please rename variables i and x and don't use a "path" dict feature which is intransparent...

File: prodsys/simulation/router.py [hunk 14]
```diff
@@ -157,10 +161,15 @@ def get_possible_resources(
         """
         # TODO: maybe make result of function in dict and save it after instantiation. 
         possible_resources = []
+        if isinstance(processing_request, request.TransportResquest):
+            processing_request.path_to_target['resource_ID'] = []
```
Comment (SebBehrendt) [inline prodsys/simulation/router.py:165]: same as above

File: prodsys/simulation/state.py [hunk 15]
```diff
@@ -368,10 +368,14 @@ def prepare_for_run(self):
     def activate_state(self):
         self.active = events.Event(self.env).succeed()
 
-    def process_state(self, target: List[float]) -> Generator:
+    def process_state(self, target: List[float], bol: bool) -> Generator:
         self.done_in = self.time_model.get_next_time(
             origin=self.resource.get_location(), target=target
         )
+        if bol == True:
```
Comment (SebBehrendt) [inline prodsys/simulation/state.py:375]: naming of bol -> give it a more descriptive name

File: prodsys/util/runner.py [hunk 16]
```diff
@@ -166,6 +171,51 @@ def initialize_simulation(self):
...
+
+            self.resource_factory = resource_factory.ResourceFactory(
+                env=self.env,
+                state_factory=self.state_factory,
+                queue_factory=self.queue_factory,
+                process_factory= self.linktransportprocess_factory
+            )
+            self.resource_factory.create_resources(self.adapter)
+
+            self.product_factory = product_factory.ProductFactory(
```
Comment (SebBehrendt) [inline prodsys/util/runner.py:192]: Here is duplicate stuff, e.g. source Factory

File: prodsys/adapters/adapter.py (hunk 2 above)
Comment (jonaskirner) [reply prodsys/adapters/adapter.py]: Commit 22

File: prodsys/express/process.py
Comment (jonaskirner) [reply prodsys/express/process.py:1]: Commit 23. I would say that capabilities still stay optional. The RequiredTransportProcess is only necessary with several LinktransportProcesses. For one linktransportprocess, we can directly use it for the product.

File: prodsys/factories/resource_factory.py
Comment (jonaskirner) [reply prodsys/factories/resource_factory.py:1]: Commit 24

File: prodsys/simulation/control.py (hunk 4 above)
Comment (jonaskirner) [reply prodsys/simulation/control.py:467]: Commit 25

File: prodsys/simulation/path_finder.py (hunk 6 above)
Comment (jonaskirner) [reply prodsys/simulation/path_finder.py:60]: Commit 26

File: prodsys/simulation/process.py (hunk 8 above)
Comment (jonaskirner) [reply prodsys/simulation/process.py:245]: Commit 27

File: prodsys/simulation/resources.py (hunk 10 above)
Comment (jonaskirner) [reply prodsys/simulation/resources.py:225]: Commit 28

File: prodsys/simulation/resources.py (hunk 12 above)
Comment (jonaskirner) [reply prodsys/simulation/resources.py]: Commit 28

File: prodsys/simulation/resources.py (hunk 11 above)
Comment (jonaskirner) [reply prodsys/simulation/resources.py:227]: Commit 29: But here it is no match with a request, but from a resource it s capability. I added the match_request in the process.

File: prodsys/simulation/router.py (hunk 13 above)
Comment (jonaskirner) [reply prodsys/simulation/router.py:89]: Commit 30

File: prodsys/simulation/state.py (hunk 15 above)
Comment (jonaskirner) [reply prodsys/simulation/state.py:375]: Commit 31

File: prodsys/simulation/router.py (hunk 14 above)
Comment (jonaskirner) [reply prodsys/simulation/router.py:165]: Commit 29

File: prodsys/simulation/request.py (hunk 9 above)
Comment (jonaskirner) [reply prodsys/simulation/request.py]: Commit 32: Possible to delete path to origin. Path__to_target needs to be added to each Transportrequest in the process.py and will then be selected by heuristic with resource in the router.py.

File: prodsys/factories/linktransportprocess_factory.py
Comment (jonaskirner) [reply prodsys/factories/linktransportprocess_factory.py:1]: Resources needs processes. But LinkTransportprocesses need ressources, sinks & sources. Thats why we said we split up in two factories and add it in the runner in the end.

**Review by SebBehrendt — APPROVED** [review]

**Pull request merged**